File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
import java.util.Arrays;

class Solution {
    public int solution(int[][] triangle) {
        
        int h = triangle.length;

        if (h == 1) return triangle[0][0];
        
        int[][] dp = new int[h][];
        for (int i = 0; i < h; i++) {
            dp[i] = new int[triangle[i].length];
        }
        
		dp[0][0] = triangle[0][0];
        
        for (int i = 1; i < h; i++) {
            
            dp[i][0] = dp[i - 1][0] + triangle[i][0];
            dp[i][i] = dp[i - 1][i - 1] + triangle[i][i];
            
            for (int j = 1; j < i; j++) {
                dp[i][j] = Math.max(dp[i-1][j-1], dp[i-1][j]) + triangle[i][j];
            }
        }
        
		return Arrays.stream(dp[h-1]).max().getAsInt();
        
    }
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    N = len(triangle)
    
    for i in range(1, N):
        for j in range(i + 1):
            n1 = triangle[i - 1][j - 1] if j > 0 else 0
            n2 = triangle[i - 1][j] if j < i else 0
            triangle[i][j] += max(n1, n2)
            
    return max(triangle[N - 1])
            
        
    